lib/modules/manager/buildpacks/extract.ts

```diff
@@ -0,0 +1,122 @@
+import is from '@sindresorhus/is';
+import { logger } from '../../../logger';
+import { regEx } from '../../../util/regex';
+import { isVersion } from '../../versioning/semver';
+import { getDep } from '../dockerfile/extract';
+import type {
+  ExtractConfig,
+  PackageDependency,
+  PackageFileContent,
+} from '../types';
+import { type ProjectDescriptor, ProjectDescriptorToml } from './schema';
+
+const dockerPrefix = regEx(/^docker:\/?\//);
+const buildpackRegistryRef = regEx(
+  /^[a-z0-9\-.]+\/[a-z0-9\-.]+(?:@(?<version>.+))?$/,
+);
+
+function isBuildpackRegistryRef(ref: string): boolean {
+  const bpRegistryMatch = buildpackRegistryRef.exec(ref);
+  if (!bpRegistryMatch) {
+    return false;
+  } else if (!bpRegistryMatch.groups?.version) {
+    return true;
+  }
+
+  return isVersion(bpRegistryMatch.groups.version);
+}
+
+function isDockerRef(ref: string): boolean {
+  const schemaMatch = regEx(/^([a-z0-9]+):\/?\//).test(ref);
+  if (
+    ref.startsWith('urn:cnb') || // buildpacks registry or builder urns
+    ref.startsWith('from=') || // builder reference
+    isBuildpackRegistryRef(ref) || // buildpack registry ID reference
+    (schemaMatch && !ref.startsWith('docker:/')) // unsupported schema
+  ) {
+    return false;
+  }
+
+  return true;
+}
```

Comment:
I think it would be more robust if this function defaulted to false and returned true when conditions are _met_, rather than defaulting to true and returning false for the four unsupported conditions you know about. Is it possible to reverse this logic for more robustness, i.e. validate that it _is_ a docker ref, not that it _isn't_?

Reply:
I am not sure about that. Because that would imply that we have a `regex` for a valid docker ref and that `regex` would be rather complex itself. And nothing "bad" will happen in case we identify something wrongly as docker ref. Nothing would be found. Thoughts?

Reply:
I'm certain that it would be more robust. But if you think it's too hard, then we can probably allow it. Please test what happens if a non-Docker ref passes this isDockerRef() function. e.g. will we attempt to look it up on Docker Hub?

Reply:
yes, that would happen. I don't like to pass them blind to docker hub when they aren't docker images as is. (eg need a different registry lookup before)

Reply:
From my understanding, we have 2 options here:

1. We handle all known non-docker use cases and the rest will be handled as docker. Which might end up trying to pull an image which is not a valid docker ref, *but* it was none of the other use cases and well and nothing but happens if it is not a docker ref, right? It just does not work.
2. We try the `regex` approach to identify a docker ref. If we miss it, it will be ignored although it would work.

I clearly prefer option 1, but it is not my call. I only think that a `regex` for a valid docker image is error-prone and gets ugly very fast imho.

Reply:
> I think it would be more robust if this function defaulted to false...

What aspect of "robustness" are you referring to?

Reply:
> > I think it would be more robust if this function defaulted to false...
> 
> What aspect of "robustness" are you referring to?

A more robust program only accepts valid input, and rejects bad input.

**Current design**

The current design defaults to `True`, and only returns `False` on obviously bad cases. This design relies on us to find, and describe, _all_ the known bad cases. Even worse, there may be bad cases we don't know about yet. :upside_down_face: 

**More robust design**

By defaulting to `False`, we can follow this logic:

1. IF the input is obviously bad, return `False` right way (fail-fast)
2. IF the input matches _all_ the expectations, rules, specifications for a Docker reference: return `True`
3. ELSE return `False` (fail-safe)

This design gives us stronger guarantees about the input.

**Deep dive**

We know of four bad cases already:

- `ref.startsWith('urn:cnb') // buildpacks registry or builder urns`
- `ref.startsWith('from=') // builder reference`
- `isBuildpackRegistryRef(ref) // buildpack registry ID reference`
- `(schemaMatch && !ref.startsWith('docker:/')) // unsupported schema`

But we can't be sure we listed _all_ possible bad cases. :upside_down_face: 

Luckily we know what a valid Docker reference looks like. The hard part is to:

- Figuring out which the conditions must be met for input to be a valid Docker reference
- Put the conditions into TypeScript code

**Proposed control flow**

Once we have the specifications, we can change the control flow to something like this:

```
// Fail fast, check for bad stuff first

// Fail on bad schemas

IF input matches:

- (schemaMatch && !ref.startsWith('docker:/')) // unsupported schema

return false

// Fail on other bad cases

IF input matches ANY of these rules:

  - `ref.startsWith('urn:cnb') || // buildpacks registry or builder urns`
  - `ref.startsWith('from=') || // builder reference`
  - `isBuildpackRegistryRef(ref) || // buildpack registry ID reference`

return false

// Check if input is a valid Docker reference

IF the input matches ALL of these rules:

  - Rule 1 to be a valid Docker reference
  - Rule 2 to be a valid Docker reference
  - Rule 3 to be a valid Docker reference
  - Rule 4 to be a valid Docker reference
  - Keep listing rules until we have the full specification for a valid Docker reference...

return true

// Finally, if none of the above matches, fail-safe

ELSE return false
```

Reply:
> yes, that would happen. I don't like to pass them blind to docker hub when they aren't docker images as is. (eg need a different registry lookup before)

https://github.com/renovatebot/renovate/blob/3a6465ee732c17c581811caac7a8e7f8320f4968/lib/modules/manager/helm-values/extract.ts#L58

@viceice We simply took inspiration from already existing code where no checks take place at all before passing it to the `getDep` function from the docker datasource. We'll try to come up with a valid regex for a positive checks.

Reply:
We flipped the detection logic and use a regex which is loosely based on the [reference implementation.](https://github.com/containers/image/blob/main/docker/reference/regexp.go)